RCTComms/transport.py

```diff
@@ -238,74 +247,163 @@ def close(self):
+            if len(ready[0]) == 1:
+                data = self.__socket.recv(bufLen)
+                return data, self.__target[0]
+            else:
+                raise TimeoutError
+        except WindowsError as e:
+            if e.winerror == 10038:
+                pass
+            else:
+                raise
 
     def send(self, data: bytes, dest=None):
         if self.__socket is None:
             raise RuntimeError()
         self.__socket.send(data)
-    
+
     def isOpen(self):
         return self.__socket is not None
 
-class RCTTCPServer(RCTAbstractTransport):
-    def __init__(self, port: int):
+class RCTTCPServer:
+    def __init__(self, port: int, connectionHandler: Callable[[RCTAbstractTransport, int], None], addr: str = ''):
+        '''
+        Creates an RCTTCPServer object to be bound to the specified port.
+        '''
         self.__log = logging.getLogger('RCT TCP Server')
         self.__port = port
         self.__socket: Optional[socket.socket] = None
-        self.__conn: Optional[socket.socket] = None
-        self.__addr: Optional[Tuple[str, int]] = None
+        self.__generatorThread: Optional[threading.Thread] = None
+        self.running = False
+        self.__hostAdr = addr
+        self.__connectionHandler = connectionHandler
+        self.__connectionIndex = 0
+        self.simList = []
 
     def open(self):
+        '''
+        Opens the server. Socket is created and generatorThread begins listening
+        for new connections.
+        '''
+        # Use printed addr in rctconfig
+        print ('Server started at {}'.format(socket.gethostbyname(socket.gethostname())))
+
         error_time = 1
         while self.__socket is None:
             try:
+                self.running = True
                 self.__socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-                self.__socket.bind(('', self.__port))
+                self.__socket.bind((self.__hostAdr, self.__port))
+                self.__log.info('Port is listening')
                 self.__socket.listen()
             except Exception as exc: # pylint: disable=broad-except
+                self.running = False
                 self.__socket = None
                 self.__log.exception('Failed to open port: %s', exc)
                 time.sleep(error_time)
                 error_time = min(2 * error_time, 10)
-        self.__log.info('Port is listening')
-        self.__conn, self.__addr = self.__socket.accept()
-        self.__log.info('Port is connected')
+        self.__generatorThread = threading.Thread(target=self.generatorLoop)
+        self.__generatorThread.start()
+
+    def generatorLoop(self):
+        '''
+        Thread to accept new connections to the server. A new
+        RCTTCPConnection object is made each time a client connects.
+        '''
+
+        while self.running:
+            try:
+                clientConn, clientAddr = self.__socket.accept()
```

Comment:
This action will block this entire loop.  `__generatorThread.join` will not cause this to exit.  As a result, `__socket.close` will never execute, and you will have hanging resources.

One option is to make generatorThread a daemon.  Other option is to add `self.running` as an Event to a selector